### DISCUSSION TITLE
PnP Search extensibility

### DISCUSSION THREAD

**Discussion by fisketryne5** [microsoft-search/pnp-modern-search#2730] [General]
Hi

We are using PnP Search as a search center for a large norwegian company. This is working great and it gives us the level of customization we need. We use it to search for employees, news, documents, quality system and more.

Until now we have used an On Premise SharePoint farm to index content on fileshares. The on prem index have over 100M items, so we cannot move the index to M365 without paying a ridiculous amout of money by using a connector. So we created a LogicApp that have access to the OnPrem SharePoint farm and modified PnP Search so that it sends the search request to this Logic App that returns a on prem search result. This works great and we can use all the functionality as before, like filtering and sorting.

But we are now looking into the possibility to create a PnP Search extension. And I am wondering if we can use a custom data source to achive the same? Does anyone know if this is possible? I have tested this, but it does no look like we get the posibility to use filtering and sorting without write the logic in the extension itself.

**Comment by patrikhellgren** [comment]
I have done something similar but with an Azure Function that "proxies" the search request. What I did was to implement a custom data source in an extension library. To this I then copied the SharePointSearchDataSource class together with dependencies from PnP Modern Search and then changed SharePointSearchDataSource and SharePointSearchService to send the search request to my Azure Function instead of the standard SharePoint API. This way we can have all features that are already implemented in the default data source like filtering etc.
This was not against SharePoint on-prem but I guess there would be no issues with this except perhaps some authentication issues depending on how you would like to authenticate.

**Reply by fisketryne5** [reply]
Excellent! It was exactly what I was looking for and it works :) 
Thank you @patrikhellgren
